### DISCUSSION TITLE
Usage of the same VIRTUAL_HOST in multiple containers

### DISCUSSION THREAD

**Discussion by pankalog** [nginx-proxy/nginx-proxy#2412] [Q&A]
Hello everyone,

I am making a project that is attempting to use nginx-proxy to dynamically assign multiple containers a single hostname, which are themselves created dynamically. We would like to maintain the look-and-feel of a "single computer system", meaning that we would like to allow multiple docker containers to appear as a single computational system, by allowing a single hostname to resolve different containers on different ports.

For example, I would like to use a single hostname (`userId.containers.example.com`) to access a container that could have an MQTT broker at port 8883, a container serving a database at port 5432, a frontend at port 80, etc.

In my case, I have set up nginx-proxy using network_mode host, which means that the requests are reaching the container. 

I took a look at `/etc/nginx/conf.d/default.conf` and I saw that the upstream for the hostname was setup properly, meaning that all 3 containers that had the same VIRTUAL_HOST were registered on that upstream, but here is the issue:

```nginx
server {
    server_name userId.containers.example.com;
    access_log /var/log/nginx/access.log vhost;
    http2 on;
    listen 80 ;
    listen 443 ssl ;
    # No certificate found for this vhost, so force nginx to emit a TLS error if
    # the client connects via https.
    ssl_ciphers aNULL;
    set $empty "";
    ssl_certificate data:$empty;
    ssl_certificate_key data:$empty;
    if ($https) {
        return 444;
    }
    location / {
        proxy_pass http://userId.containers.example.com;
        set $upstream_keepalive false;
    }
}
```

This shows that only ports 80 and 443 are being listened to. 

I would like to figure out a way to provide a 1-to-1 mapping of container ports to the VIRTUAL_HOST. so if the following ports are exposed:

container1:80
container2:5432
container2:3000
container3:8883

I would like them to be accessible from userId.containers.example.com:80,5432,3000,8883, etc. respectively.

Is there a way to do this using nginx-proxy or is this not really the correct use-case for this?

Thanks for your help, and thank you for maintaining!

**Comment by buchdag** [comment]
@pankalog I'm not 100% sure I got what you want to do but if it is this:

- `userId.containers.example.com:80` -> `container1:80`
- `userId.containers.example.com:5432` -> `container2:5432`
- `userId.containers.example.com:3000` -> `container2:3000`
- `userId.containers.example.com:8883` -> `container3:8883`

Then no, nginx-proxy with its current feature set won't allow you to do this.

What is possible is either this (assuming port 80 on left side):
- `userId.containers.example.com` -> `container1:80`
- `userId.service2.containers.example.com` -> `container2:5432`
- `userId.service3.containers.example.com` -> `container2:3000`
- `userId.service4.containers.example.com` -> `container3:8883`

Or this:
- `userId.containers.example.com` -> `container1:80`
- `userId.containers.example.com/service2` -> `container2:5432`
- `userId.containers.example.com/service3` -> `container2:3000`
- `userId.containers.example.com/service4` -> `container3:8883`

Or any mix of both.

What confuses me is that you mention running your container using network_mode host (or is the proxy alone ?). If that's the case, your container are already directly reachable on their respective port on the host and you don't need nginx-proxy at all.

**Reply by pankalog** [reply]
Hey @buchdag ,
Yes, the first thing you mentioned is exactly what I want to do. 

Is there any way for me to do this, with any tool? I'm starting to think that it's not "physically" possible, due to the IP protocol itself.

Thank you so much!

**Reply by buchdag** [reply]
I think I might have misunderstood something 🤔 

Do you have multiple containers sets running on each host for different userIds, ie:
- `user1.containers.example.com:80` -> `container1:80`
- `user1.containers.example.com:5432` -> `container2:5432`
- `user1.containers.example.com:3000` -> `container2:3000`
- `user1.containers.example.com:8883` -> `container3:8883`
- `user2.containers.example.com:80` -> `container4:80`
- `user2.containers.example.com:5432` -> `container5:5432`
- `user2.containers.example.com:3000` -> `container5:3000`
- `user2.containers.example.com:8883` -> `container6:8883`

...etc ? Or is it one host -> one userId ?

**Reply by pankalog** [reply]
Everything would be hosted in a single host, in the way you said

**Reply by buchdag** [reply]
> I'm starting to think that it's not "physically" possible, due to the IP protocol itself.

That depends on what service is being served on those ports. If it's HTTP then it could theoretically work without issue, but nginx-proxy itself does not support this (nothing prevents you from customising the `nginx.tmpl` you're using, but I'll have to admit that might not be the easiest task 😒).

If its not HTTP that's a whole other issue, you'll have to use [`ngx_stream_core_module`](https://nginx.org/en/docs/stream/ngx_stream_core_module.html) and nginx-proxy does not have any builtin support for automatic proxying using this module, so you'll be pretty much writing a new `nginx.tmpl` on your own.

[That also seems theoretically possible though.](https://blog.nashcom.de/nashcomblog.nsf/dx/nginx-tcp-stream-with-sni-support.-more-than-helpful-for-lab-environments.htm?opendocument&comments)

Edit: also, the obvious suggestion would be to check if [Traefik](https://github.com/traefik/traefik) is able to do this.

**Reply by pankalog** [reply]
I've tried really hard, but Traefik was not something that I could use. The issue with this is that you essentially have to bind all ports on that host beforehand, and make sure that the redirects happen properly (Traefik does not have a mechanism that allows dynamic/thousands of entrypoints). I thought of handwriting something like that using Rust, but it seems that it's too much work for something like this. It would be cool tho, for sure, but it doesn't seem like something worth spending time on. 

An idea I've had that I think could work, is updating a local DNS server that runs as a Docker container, and then getting that to deal with redirects by inserting/removing A records. Do you know how something like this would work? 

Thanks for replying!
